### DISCUSSION TITLE
Clarifying intended differences between "MinimumViableSpacing" and "Spaces"

### DISCUSSION THREAD

**Discussion by aegershman** [openrewrite/rewrite#271] [Q&A] [answered]
I'd like to ask this to clarify my own understanding & possibly help anyone else searching for this whose wondering the same thing:

**What's the difference in _purpose_ between `MinimumViableSpacing{Visitor}` and `Spaces{Visitor}`** in `rewrite-java`?

- I'm under the impression that `MinimumViableSpacing` deals with-- as the name implies-- the minimum amount of spacing that's required in order for the tree to be valid, compilable Java code. E.g., given `return "foo";`, it's not just "stylistic" to have a space between `return` and `"foo"`, because `return"foo";` is not valid. There must be a minimum of 1 space between the `return` keyword and `"foo"`. The "practical" use-case we rely on `MinimumViableSpacing` for is cases where we _generate_ code output and don't want to have to manually manage spacing ourselves.
- Comparatively, `Spaces` deals with _styling_ of spacing on tree elements which should be otherwise be valid, but configurable as a preference.

Is this a generally-fair assessment? Thoughts/feedback welcome.

**Comment by jkschneider** [comment, marked as answer]
That's exactly right!
